src/task.c

```diff
@@ -561,7 +561,7 @@ static void JL_NORETURN throw_internal(jl_task_t *ct, jl_value_t *exception JL_M
     ptls->io_wait = 0;
     // @time needs its compile timer disabled on error,
     // and cannot use a try-finally as it would break scope for assignments
-    jl_measure_compile_time[ptls->tid] = 0;
```

Comment:
Ah, actually, @janrous-rai raises a really great point here.

> Even if this is used "correctly" as an alternative exit point from the `@time` call I am not sure if it is only guaranteed to occur in exactly that context, i.e. whether we can guarantee that we won't get the `jl_measure_compile_time` counter decremented in other scenarios.

Thank you for the careful review! You're totally right. As currently implemented, even if the user hasn't called `@time`, and then we throw an exception, this will decrement the counter, and then we'll _start_ measuring compilation time because `if (-1)` is considered `true` in C! 😂 And then when the user calls `jl_cumulative_compile_time_before`, we'll _stop_ measuring. 😬 
Thanks for catching this!

If our only goal was to support reentrant calls to `@time`, I think we could address this by using an atomic compare-and-decrement that only decrements this **if it's already above 0**. 

However, we _also_ would like to support the use-case identified at RelationalAI, where we just want to turn this on all the time, and _leave it on_, and in that case we obviously don't want to disable it whenever we throw an exception!

🤔 This requires more thinking. 😕
Thanks for identifying it.

Reply:
My other PR deals with the other use-case and is somewhat simpler as it doesn't need to deal with re-entrancy (which is conceptually hard) and because it is triggered on demand, I think it's okay if the compilation costs incurred by exception handling is included in the timing (we care more about the rough-estimates and excluding or including the exception handling compilation costs will likely be miniscule in the overall big picture).

Reply:
Hah, bummer, another reason that this original mechanism is broken is that it must mean that a **handled** try/catch would already disable compilation time measurement! 

And yeah, indeed, seems like it:
```julia
julia> @time begin
           # The following exception disables compilation time measurement
           try throw("ha HAH!") catch end
           @eval module M ; f(x) = 2*x ; end
           @eval M.f(2)
       end
  0.006116 seconds (2.17 k allocations: 156.874 KiB)
4
```
vs
```julia
julia> @time begin
           @eval module M ; f(x) = 2*x ; end
           @eval M.f(2)
       end
WARNING: replacing module M.
  0.008430 seconds (1.20 k allocations: 84.593 KiB, 53.77% compilation time)
4
```

Oops!

Reply:
Oh, it sounds like this is already reported and understood, so i think we can leave this alone in this PR. :)

Reply:
Here's the relevant discussions about this existing bug where exceptions disable the compilation measurement:

https://github.com/JuliaLang/julia/pull/38915
https://github.com/JuliaLang/julia/pull/39138
https://github.com/JuliaLang/julia/pull/39217

Reply:
As mentioned in another comment (https://github.com/JuliaLang/julia/pull/41733/#issuecomment-901368198), we're going to have to tackle this exception throwing issue in a separate PR.

This PR is an improvement, and we don't want to hang it up.

Thanks for pointing this out!